samples/touchtest/touchtest.cpp

```diff
@@ -0,0 +1,209 @@
+    void OnMouseMove(wxMouseEvent& event);
+    void OnMouseUp(wxMouseEvent& event);
+
+    void OnQuit(wxCommandEvent& event);
+
+DECLARE_EVENT_TABLE()
+};
+
+IMPLEMENT_APP(MyApp)
+
+bool MyApp::OnInit()
+{
+    if ( !wxApp::OnInit() )
+        return false;
+
+    wxFrame* frame = new MyFrame(NULL, wxT("Multi-touch Test"), wxDefaultPosition,
+        wxSize(500, 400), wxDEFAULT_FRAME_STYLE | wxHSCROLL | wxVSCROLL);
+    SetTopWindow(frame);
+    frame->CenterOnScreen();
+    frame->Show(true);
+
+    return true;
+}
+
+wxBEGIN_EVENT_TABLE(MyFrame, wxFrame)
+    EVT_MENU(wxID_EXIT, MyFrame::OnQuit)
+    EVT_TOUCH_BEGIN(MyFrame::OnTouchBegin)
+    EVT_TOUCH_MOVE(MyFrame::OnTouchMove)
+    EVT_TOUCH_END(MyFrame::OnTouchEnd)
+    EVT_TOUCH_CANCEL(MyFrame::OnTouchEnd)
+    EVT_LEFT_DOWN(MyFrame::OnMouseDown)
+    EVT_MOTION(MyFrame::OnMouseMove)
+    EVT_LEFT_UP(MyFrame::OnMouseUp)
+wxEND_EVENT_TABLE()
+
+MyFrame::MyFrame(wxFrame *parent, const wxString& title, const wxPoint& pos,
+    const wxSize& size, const long style)
+    : wxFrame(parent, wxID_ANY, title, pos, size, style),
+    m_MouseId(&m_MouseId)
+{
+    m_TouchPoints[0].pen = wxPen(*wxBLACK, 2);
+    m_TouchPoints[1].pen = wxPen(*wxBLUE, 2);
+    m_TouchPoints[2].pen = wxPen(*wxCYAN, 2);
+    m_TouchPoints[3].pen = wxPen(*wxGREEN, 2);
+    m_TouchPoints[4].pen = wxPen(*wxYELLOW, 2);
+    m_TouchPoints[5].pen = wxPen(*wxRED, 2);
+    m_TouchPoints[6].pen = wxPen(*wxLIGHT_GREY, 2);
+    m_TouchPoints[7].pen = wxPen(*wxBLACK, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[8].pen = wxPen(*wxBLUE, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[9].pen = wxPen(*wxCYAN, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[10].pen = wxPen(*wxGREEN, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[11].pen = wxPen(*wxRED, 2, wxPENSTYLE_DOT_DASH);
+    m_TouchPoints[12].pen = wxPen(*wxBLACK, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[13].pen = wxPen(*wxYELLOW, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[14].pen = wxPen(*wxBLUE, 2, wxPENSTYLE_DOT);
+    m_TouchPoints[15].pen = wxPen(*wxRED, 2, wxPENSTYLE_DOT);
+
+    wxMenu *file_menu = new wxMenu;
+
+    file_menu->Append(wxID_EXIT, wxT("&Exit"));
+
+    wxMenuBar *menu_bar = new wxMenuBar;
+
+    menu_bar->Append(file_menu, wxT("&File"));
+    SetMenuBar(menu_bar);
+
+    EnableTouchEvents(wxTOUCH_RAW_EVENTS);
+}
+
+void MyFrame::OnQuit(wxCommandEvent& WXUNUSED(event))
+{
+    Close(true);
+}
+
+void MyFrame::DrawStart(const wxTouchPointId& id, wxPoint pos)
+{
+    int idx = -1;
+    for(unsigned i = 0; i < TOUCH_POINTS; i++)
+        if (m_TouchPoints[i].id == id)
+            idx = i;
+    if (idx == -1)
+        for(unsigned i = 0; i < TOUCH_POINTS; i++)
+            if (!m_TouchPoints[i].id.IsOk())
+	    {
+                idx = i;
+		break;
+	    }
+    if (idx == -1)
+        return;
+    m_TouchPoints[idx].id = id;
+    m_TouchPoints[idx].last = pos;
+}
+
+void MyFrame::DrawUpdate(const wxTouchPointId& id, wxPoint pos)
+{
+    int idx = -1;
```

Comment:
This repeated snippet of code would become more clear if it were extracted into some `FindIndexOfTouchId()` function.